File: src/main/java/com/lessmarkup/framework/helpers/PropertyDescriptor.java
package com.lessmarkup.framework.helpers;

import com.lessmarkup.interfaces.exceptions.CommonException;
import java.lang.annotation.Annotation;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public class PropertyDescriptor {
    private final Method getter;
    private final Method setter;
    private final String name;
    private final Class<?> type;
    
    public PropertyDescriptor(String name, Method getter, Method setter, Class<?> type) {
        this.getter = getter;
        this.setter = setter;
        this.name = name;
        this.type = type;
    }

    public Object getValue(Object thisObject) {
        try {
            return this.getter.invoke(thisObject);
        } catch (IllegalAccessException | IllegalArgumentException ex) {
            throw new CommonException(ex);
        } catch (InvocationTargetException ex) {
            throw new CommonException(ex.getTargetException());
        }
    }
    
    public void setValue(Object thisObject, Object newValue) {
        try {
            this.setter.invoke(thisObject, newValue);
        } catch (IllegalAccessException | IllegalArgumentException ex) {
            throw new CommonException(ex);
        } catch (InvocationTargetException ex) {
            throw new CommonException(ex.getTargetException());
        }
    }

    public String getName() {
        return name;
    }

    public Class<?> getType() {
        return type;
    }

    public <T extends Annotation> T getAnnotation(Class<T> annotationClass) {
        T ret = this.getter.getAnnotation(annotationClass);
        if (ret != null) {
            return ret;
        }
        return this.setter.getAnnotation(annotationClass);
    }
}

File: src/main/scala/com/lessmarkup/framework/helpers/PropertyDescriptor.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.framework.helpers

import java.lang.annotation.Annotation
import java.lang.reflect.{InvocationTargetException, Method}

import com.lessmarkup.interfaces.exceptions.CommonException

class PropertyDescriptor(name: String, getter: Method, setter: Method, dataType: Class[_]) {

  def getValue(thisObject: AnyRef): AnyRef = {
    try {
      getter.invoke(thisObject)
    }
    catch {
      case ex: InvocationTargetException => throw new CommonException(ex.getTargetException)
      case ex: Any => throw new CommonException(ex)
    }
  }

  def setValue(thisObject: AnyRef, newValue: AnyRef) {
    try {
      setter.invoke(thisObject, newValue)
    }
    catch {
      case ex: InvocationTargetException => throw new CommonException(ex.getTargetException)
      case ex: Any => throw new CommonException(ex)
    }
  }

  def getName: String = {
    name
  }

  def getType: Class[_] = dataType

  def getAnnotation[T <: Annotation](annotationClass: Class[T]): T = {
    val ret: T = getter.getAnnotation(annotationClass)
    if (ret != null) {
      return ret
    }
    setter.getAnnotation(annotationClass)
  }
}